Expose API that can be used to fetch rest stops.

abhishek1508 opened the pull request:
### Description
<!--
Include issue references (e.g., fixes [#issue](link))
Include necessary implementation details (e.g. I opted to use this algorithm because ... and test it in this way ...).
-->
Refs: https://github.com/mapbox/navigation-sdks/issues/1840

Even though the PR adds API that allows users to fetch rest stops, this PR will serve as the base for exposing other road objects down the line; eg: `tunnel`, `tolls`, `bridges`, `incidents`, `railwayCrossing` etc.

The code introduced in this PR allows users to fetch rest stops when traveling on a route that either:
- goes to the rest stop,
- passes through the rest stop.

In addition, the code also allows users to fetch rest stop during 
- **route preview mode using `navigationRoute`:** to render all rest stops on a route line at appropriate location.
- **active navigation mode using `routeProgress`:** to render rest stop as the user travels.
- **free drive mode using `EHorizon`:** to render rest stop as the user is exploring.

When implementing the code, I found that when returning `restStop` for `active navigation` and `free drive` mode, I am making use of `UpcomingRoadObject` of the type `RestStop`

```
fun getRestStop(upcomingRoadObjects: List<UpcomingRoadObjects>): RestStop
```

`RestStop` is simply derived from this `upcomingRoadObject` and returned to the caller of the API

On the other hand, when returning `restStop` for `route preview` mode, I am making use of `navigationRoute`

```
fun getRestStops(navigationRoute: NavigationRoute): Map<RouteLeg, List<RestStop>>
```

But if you look deep you will see that `RestStop` in the SDK has an `internal constructor` and hence I cannot create this object using `navigationRoute`. Hence, I created a sealed class as following:

```
sealed class MapboxRestStop {

    data class FromRoute(
        val name: String?,
        val type: String?,
        val amenities: List<Amenity>?,
        val location: Point
    ): MapboxRestStop()

    data class FromEHorizonOrRouteProgress(
        val restStop: RestStop,
        val distanceToStart: Double
    ): MapboxRestStop()
}
```

and I made both the API(s) return `MapboxRestStop` as following:

```
fun getRestStop(upcomingRoadObjects: List<UpcomingRoadObjects>): MapboxRestStop?

fun getRestStops(navigationRoute: NavigationRoute): Map<RouteLeg, List<MapboxRestStop>>
```

I am open to hear more ideas if anyone has better one.

@VysotskiVadim @RingerJK @LukasPaczos 

<!--
---------- CHECKLIST ----------
1. Add related labels (`bug`, `feature`, `new API(s)`, `SEMVER-MAJOR`, `needs-backporting`, etc.).
1. Adda a changelog entry under `Unreleased` tag or a `skip changelog` label if not applicable.
1. Update progress status on the project board.
1. Request a review from the team, if not a draft.
1. Add targeted milestone, when applicable.
1. Create ticket tracking addition of public documentation pages entry, when applicable.
-->

dzinad commented:
How is this API supposed to be used. It would be nice to see an example. 
I see it like this:
`(MapboxRoadObjectApi().getRestStop(upcomingRoadObjects) as FromEHorizonOrRouteProgress?).restStop`
`(MapboxRoadObjectApi().getRestStop(navigationRoute)[0] as List<FromRoute>)[0].amenities`
The first thing that comes to mind is that we know which subclass will be returned for which method, so why not specify it in the API?
Also, have you thought about introducing a single class for both methods? Name, type, amenities and location exist in both cases (although location is a bit different). The only other difference is the distanceFromStart property. But is it really a RestStop? We could return a list of Pairs, like you do with a map for stops from navigationRoute.
Another option is to create helper method in the right module so that we can return RestStop. But I'm not sure why the constructor was made internal. Do we really want to forbid creating this object?

RingerJK commented:
- I agree with @dzinad, would be great to have a look at how this API would be used. 
- From BMW's request, I got that they want to show users all upcoming `RoadStops`, so probably, API should look like: 
```
fun getRestStop(upcomingRoadObjects: List<UpcomingRoadObject>): List<MapboxRestStop>
```
- just my preferences, I would replace static functions by extensions, it would look like filtes
```
fun List<UpcomingRoadObject>.restStops(): List<FreeDriveRestStop>
fun List<UpcomingRoadObject>.tunnels(): List<FreeDriveTunnels>
fun NavigationRoute.restStops(): List<RoadRestStops>
```
- `sealed class MapboxRestStop` is not extended, change it to empty `interface RoadStopsInterface` to just mark `RoadStops`
- data classes are not great for public API, they're not extended
- if RoadStops have the same properties, can be used facade aka
```
internal class RestStopsBase(val name: String)
class RoadRestStop constructor (OTHER_FIELDS, internal val base: RestStopsBase) {
  val name get() = base.name
  // builder
  fun build(): RoadRestStop{  
    // other fields
    return RoadRestStop(OTHER_FIELDS, RestStopsBase(name))
  }
}
```
they don't inherit and looks like it's fine as they are used in separate sessions. If need to add common fields it will be added to `Base` once

abhishek1508 commented:
> Also, have you thought about introducing a single class for both methods? Name, type, amenities and location exist in both cases (although location is a bit different)

```
class RestStop internal constructor(
    id: String,
    @RestStopType.Type val restStopType: Int,
    val name: String?,
    val amenities: List<Amenity>?,
    length: Double?,
    location: RoadObjectLocation,
    @RoadObjectProvider.Type provider: String,
    nativeRoadObject: com.mapbox.navigator.RoadObject,
) : RoadObject(id, RoadObjectType.REST_STOP, length, location, provider, nativeRoadObject)
```
I see quite many useful information that might be required when retrieving this object using EHorizon. I am afraid if we create a new single class with `name`, `type`, `amenities` etc, customers are gonna loose information. cc @dzinad 

> just my preferences, I would replace static functions by extensions, it would look like filtes

I like this idea

> they don't inherit and looks like it's fine as they are used in separate sessions. If need to add common fields it will be added to Base once

I don't like the idea of separating rest stop based on the state of navigation as shown in the example `FreeDriveRestStop`, `RoadRestStop` etc.

I have changed the implementation as follows:

I introduced `RestStopOrigin`
```
internal sealed interface RestStopOrigin {

    object Unknown: RestStopOrigin

    object NavigationRoute: RestStopOrigin

    object UpcomingRoadObjects: RestStopOrigin
}
```
Followed by `RestStopResult`
```
sealed interface RestStopResult {
    val name: String?

    data class FromIntersection internal constructor(
        override val name: String?,
        val location: Point,
        val type: String?,
        val amenities: List<Amenity>?
    ): RestStopResult

    data class FromRoadObject internal constructor(
        override val name: String?,
        val restStop: RestStop,
        val distanceToStart: Double?
    ): RestStopResult
}
```
See how I have made the `data class` to be constructed as `internal`. So this will not result in breaking changes.

Lastly

```
class RestStopData private constructor(
    val restStopWithRoadObjects: List<RestStopResult.FromRoadObject> = emptyList(),
    val restStopWithIntersection: List<RestStopResult.FromIntersection> = emptyList(),
) {

    class Builder {
        private var restStopOrigin: RestStopOrigin = RestStopOrigin.Unknown
        private var withRoadObjects: List<UpcomingRoadObject> = emptyList()
        private var withStepIntersection: List<StepIntersection>? = emptyList()

        fun withStepIntersection(stepIntersection: List<StepIntersection>?): Builder =
            apply {
                this.restStopOrigin = RestStopOrigin.NavigationRoute
                this.withStepIntersection = stepIntersection
            }

        fun withUpcomingRoadObjects(withUpcomingRoadObjects: List<UpcomingRoadObject>): Builder =
            apply {
                this.restStopOrigin = RestStopOrigin.UpcomingRoadObjects
                this.withRoadObjects = withUpcomingRoadObjects
            }

        fun build(): RestStopData {
            return when (restStopOrigin) {
                RestStopOrigin.NavigationRoute -> {
                    // create `List<RestStopResult.FromIntersection>` based on `stepIntersection` 
                    // return RestStopData(restStopWithIntersection = result)
                }
                RestStopOrigin.UpcomingRoadObjects -> {
                    // create `List<RestStopResult.FromRoadObject>` based on `upcomingRoadObject` 
                    // return RestStopData(restStopWithRoadObjects = result)
                }
                else -> {
                    throw IllegalStateException("$restStopOrigin cannot be Unknown")
                }
            }
        }
    }
```

To use the above, I have exposed API(s)

```
fun List<UpcomingRoadObject>.getAllRestStops(): List<RestStopResult.FromRoadObject>()
fun List<UpcomingRoadObject>.getFirstUpcomingRoadObject(): RestStopResult.FromRoadObject?()
fun NavigationRoute.getAllRestStops(): Map<RouteLeg, List<RestStopResult.FromIntersection>>()
```

cc @RingerJK @dzinad

abhishek1508 commented:
@RingerJK @dzinad 
Based on the suggestions above, I made some changes to the implementation. I have added a qa activity `RestStopActivity` that can be used to test the working of the API(s). To see the activity in action, make sure you filter the log statements by `RestStopActivity` and you should see appropriate logs. Make sure you add a new access token `mapbox_access_token_sapa` to `mapbox_access_token.xml` with the correct access token for the activity to work.

cafesilencio commented:
> Make sure you add a new access token mapbox_access_token_sapa to mapbox_access_token.xml with the correct access token for the activity to work.

If this is going to be needed to successfully use the activity in the QA can you add this to the activity notes accessed via the "info" label in the app.?

cafesilencio commented:
> make sure you filter the log statements by RestStopActivity and you should see appropriate logs

This doesn't seem like the best usage of the QA app.  How about something visual even if it's just a toast.  If the addition to the QA app. is really just an example then it should be put in the examples repo rather than the QA app.

dzinad commented:
> Based on the suggestions above, I made some changes to the implementation. I have added a qa activity RestStopActivity that can be used to test the working of the API(s).

It would also be nice to add the example of `NavigationRoute#getAllRestStops` usage. When you add visuals and so on, for now it's fine.

RingerJK commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/roadobject/reststop/RestStop.kt:
```diff
@@ -10,13 +10,15 @@ import com.mapbox.navigation.base.trip.model.roadobject.location.RoadObjectLocat
     id: String,
     @RestStopType.Type val restStopType: Int,
     val name: String?,
+    val amenities: List<Amenity>?,
```
add field `amenities` to `equals`, `hasCode`, `toString`

RingerJK replied:
```suggestion
    val amenities: List<Amenity>,
```

abhishek1508 replied:
This will be nullable as `amenities` is optional in directions response. @RingerJK

RingerJK commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/roadobject/RoadObjectEx.kt:
```diff
@@ -0,0 +1,65 @@
+ * - [EHorizon] during Free Drive or
+ * - [RouteProgress] during Active Navigation.
+ */
+fun List<UpcomingRoadObject>.getFirstUpcomingRoadObject(): RestStopFromRoadObject? {
```
```suggestion
fun List<UpcomingRoadObject>.getFirstUpcomingRestStop(): RestStopFromRoadObject? {
```

RingerJK commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/roadobject/RoadObjectEx.kt:
```diff
@@ -0,0 +1,65 @@
+fun List<UpcomingRoadObject>.getFirstUpcomingRoadObject(): RestStopFromRoadObject? {
+    val restStops = this.getAllRestStops()
+    if (restStops.isNotEmpty()) {
+        val distanceToStart = restStops.first().distanceToStart
```
should try to find first rest stop with nonNull `distanceToStart`?

dzinad replied:
Maybe even appropriate distanceToStart, not just nonNull? I mean so that the if below returns true.

abhishek1508 replied:
`distanceToStart` has to be `null`. This the documentation of `distanceToStart` from `nav-native`
```
@param distanceToStart remaining distance to the start of the object.
 * This value will be negative after passing the start of the object and until we cross the finish
 * point of the [RoadObject]s geometry for objects that are on the actively navigated route,
 * but it will be zero for [EHorizon] objects. It will be null if couldn't be determined.
```

RingerJK commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/roadobject/reststop/RestStopFromIntersection.kt:
```diff
@@ -0,0 +1,127 @@
+ * @param amenities facilities available in the rest stop. See [Amenity]
+ */
+class RestStopFromIntersection private constructor(
+    val location: Point?,
```
`StrepIntersection` location is NonNull, should be good to set it as well
```suggestion
    val location: Point,
```

RingerJK commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@ Mapbox welcomes participation and contributions from everyone.
 
 ## Unreleased
 #### Features
+Added extension function `RoadObjectEx` to fetch `SAPA` amenities based on `RouteProgress` and `NavigationRoute`. [#6007](https://github.com/mapbox/mapbox-navigation-android/pull/6007)
```
let's extend changelog for each fun (they're only 3)

RingerJK commented on libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/roadobject/reststop/Amenity.kt:
```diff
@@ -0,0 +1,50 @@
+     */
+    override fun hashCode(): Int {
+        var result = type.hashCode()
+        result = 31 * result + (name?.hashCode() ?: 0)
+        result = 31 * result + (brand?.hashCode() ?: 0)
```
there and below in other pojo

dzinad commented on libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/roadobject/reststop/RestStopFactoryTest.kt:
```diff
@@ -0,0 +1,105 @@
+        )
+
+        assertEquals(1, actual.size)
+        assertEquals(mockRestStop.name, actual.first().restStop?.name)
```
Maybe use `!!` here and below?

RingerJK commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/roadobject/RoadObjectEx.kt:
```diff
@@ -0,0 +1,65 @@
+ * This is recommended to be used with:
+ * - [NavigationRoute] during Route Preview
+ */
+fun NavigationRoute.getAllRestStops(): Map<RouteLeg, List<RestStopFromIntersection>> {
```
it would be hard to exted later if needed, what if we add wrapper like
```kotlin
class RouteRestsStops() {
  /**route id*/
  routeId: Int
  /**index of leg and rest stops*/
  val legsRestStops: Map<Int, List<RestStopFromIntersection>
}
```

Also, can be extended for Incidents and other things 
` this PR will serve as the base for exposing other road objects down the line`
```
class RouteExtra<T> -> class RouteExtra<RestStopFromIntersection>
class RouteExtra<T> -> class RouteExtra<incidents>
```

dzinad commented on libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/roadobject/RoadObjectExTest.kt:
```diff
@@ -0,0 +1,141 @@
+        val upcomingRoadObjects = listOf<UpcomingRoadObject>(mockk(), mockk())
+        val mockRestStops = listOf<RestStopFromRoadObject>(
+            mockk {
+                every { distanceToStart } returns null
```
You can also check if distanceToStart is zero and less than zero.

dzinad commented on libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/roadobject/RoadObjectExTest.kt:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    @Test
+    fun `no rest stops returned when route leg step is null`() {
```
You can also check that other legs are not affected (only the one with null steps is ignored).

dzinad commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/roadobject/reststop/RestStopFactory.kt:
```diff
@@ -0,0 +1,49 @@
+        return if (!stepIntersections.isNullOrEmpty()) {
+            val restStops = stepIntersections.filter { stepIntersection ->
+                stepIntersection.restStop() != null
+            }.map { intersection ->
```
```
            val restStops = upcomingRoadObjects.mapNotNull { upcomingRoadObject ->
                (upcomingRoadObject.roadObject as? RestStop)?.let { restStop ->
                    RestStopFromRoadObject
                        .Builder()
                        .restStop(restStop)
                        .distanceToStart(upcomingRoadObject.distanceToStart)
                        .build()
                }
            }
```

abhishek1508 commented:
> https://github.com/mapbox/mapbox-navigation-android/pull/6007#discussion_r915012424

This is probably a bug on nav native side. Amenities are only available for certain data providers and geography. This will be null in many situations. Will raise it up nav native. cc @RingerJK

abhishek1508 commented on CHANGELOG.md:
```diff
@@ -4,6 +4,9 @@ Mapbox welcomes participation and contributions from everyone.
 
 ## Unreleased
 #### Features
```
Done.

abhishek1508 commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/roadobject/reststop/RestStopFromRoadObject.kt:
```diff
@@ -0,0 +1,82 @@
+ */
+class RestStopFromRoadObject private constructor(
+    val restStop: RestStop,
+    val distanceToStart: Double?
```
Done.

RingerJK reviewed (approved):
LGTM

LukasPaczos commented on libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/roadobject/RoadObjectEx.kt:
```diff
@@ -0,0 +1,69 @@
+ * This is recommended to be used with:
+ * - [NavigationRoute] during Route Preview
+ */
+fun NavigationRoute.getAllRestStops(): RoadObjectExtra<RestStopFromIntersection> {
+    val legIndexToRestStops = ifNonNull(this.directionsRoute.legs()) { routeLegs ->
+        val legToRestStops = hashMapOf<Int, List<RestStopFromIntersection>>()
+        routeLegs.forEachIndexed { legIndex, routeLeg ->
+            legToRestStops[legIndex] = ifNonNull(routeLeg.steps()) { steps ->
+                val restStops = mutableListOf<RestStopFromIntersection>()
+                steps.forEach { legStep ->
+                    restStops.addAll(
+                        RestStopFactory.createWithStepIntersection(
+                            legStep.intersections()
+                        )
+                    )
+                }
+                restStops
+            } ?: emptyList()
+        }
+        legToRestStops
+    } ?: emptyMap()
+    return RoadObjectExtra(this.id, legIndexToRestStops)
+}
```
Do we need to introduce a new object type here to represent the rest stop data in the form of `RoadObjectExtra<RestStopFromIntersection>`?

Couldn't we reuse the the existing mapping functions and the fact that the `NavigationRoute` has access to the native peer? For example (we could of course optimize the way we parse to avoid work duplication):
```kotlin
fun NavigationRoute.getAllRestStops(): List<UpcomingRoadObject> {
    return this.nativeRoute().routeInfo.alerts.toUpcomingRoadObjects().filter {
        it.roadObject is RestStop
    }
}
```

abhishek1508 replied:
I still think we should have a `Map<LegIndex, List<RestStop>>` otherwise there is no way to know the rest stops fetched using `NavigationRoute` belongs to which `RouteLeg` in case there is more than 1 `RouteLeg` in the `DirectionsRoute`. Following the above, the function signature should still be

```
fun NavigationRoute.getAllRestStops(): RoadObjectExtra<UpcomingRoadObject>
```

cc @LukasPaczos
